### PR TITLE
feat: update watermark offset bucket implementation

### PR DESCRIPTION
**Pull request by chromevoid**
fix: #301

### CODE REVIEW

File: pkg/watermark/otbucket/otbucket.go [hunk 1]
```diff
@@ -0,0 +1,32 @@
+package otbucket
```
Comment (chromevoid) [inline pkg/watermark/otbucket/otbucket.go:1]: open to better names..

File: config/advanced-install/namespaced-controller-wo-crds.yaml [hunk 2]
```diff
@@ -144,7 +144,7 @@ data:
             maxAckPending: 20000
           otBucket:
             maxValueSize: 0
-            history: 1
+            history: 5
```
Comment (chromevoid) [inline config/advanced-install/namespaced-controller-wo-crds.yaml:147]: Updated all otBucket.history from 1 to 5...
Do we need to update all of these yaml file configs?.. not sure how this part works.

File: pkg/watermark/processor/entity.go [hunk 3]
```diff
@@ -33,47 +30,25 @@ func (w Watermark) Before(t time.Time) bool {
...
 // monotonically increase.
 type ProcessorEntitier interface {
 	GetID() string
-	BuildOTWatcherKey(Watermark) string
-	ParseOTWatcherKey(string) (int64, bool, error)
+	BuildOTWatcherKey() string
 }
 
 // ProcessorEntity implements ProcessorEntitier.
 type ProcessorEntity struct {
```
Comment (whynowy) [inline pkg/watermark/processor/entity.go:41]: `processorEntity`?

File: pkg/watermark/processor/entity.go [hunk 4]
```diff
@@ -33,47 +30,25 @@ func (w Watermark) Before(t time.Time) bool {
...
-
 // NewProcessorEntity returns a new `ProcessorEntity`.
-func NewProcessorEntity(name string, inputOpts ...EntityOption) *ProcessorEntity {
-	opts := &entityOptions{
-		keySeparator: _defaultKeySeparator,
-	}
-	for _, opt := range inputOpts {
-		opt(opts)
-	}
+func NewProcessorEntity(name string) *ProcessorEntity {
```
Comment (whynowy) [inline pkg/watermark/processor/entity.go:49]: Return `ProcessorEntitier`?

File: pkg/watermark/publish/publisher.go [hunk 5]
```diff
@@ -124,26 +127,21 @@ func (p *publish) PublishWatermark(wm processor.Watermark, offset isb.Offset) {
...
-		if latestWatermark < epoch {
-			latestWatermark = epoch
-		}
+	var (
+		timeWatermark = time.UnixMilli(math.MinInt64)
+		key           = p.entity.BuildOTWatcherKey()
+	)
+	byteValue, err := p.otStore.GetValue(p.ctx, key)
+	if err != nil {
+		p.log.Errorw("Unable to load latest watermark from ot store (failed to get value from ot store)", zap.String("OT", p.otStore.GetStoreName()), zap.String("processor entity", p.entity.GetID()), zap.Error(err))
```
Comment (whynowy) [inline pkg/watermark/publish/publisher.go:136]: log key in one word.

File: pkg/watermark/publish/publisher.go [hunk 6]
```diff
@@ -124,26 +127,21 @@ func (p *publish) PublishWatermark(wm processor.Watermark, offset isb.Offset) {
...
+		return processor.Watermark(timeWatermark)
+	}
+	otValue, err := otbucket.DecodeToOTValue(byteValue)
+	if err != nil {
+		p.log.Errorw("Unable to load latest watermark from ot store (failed to decode ot value)", zap.String("OT", p.otStore.GetStoreName()), zap.String("processor entity", p.entity.GetID()), zap.Error(err))
```
Comment (whynowy) [inline pkg/watermark/publish/publisher.go:141]: same here.

File: config/advanced-install/namespaced-controller-wo-crds.yaml (hunk 2 above)
Comment (chromevoid) [reply config/advanced-install/namespaced-controller-wo-crds.yaml:147]: `docs/numaflow-controller-config.yaml` this one needs a separate update

File: pkg/watermark/otbucket/otbucket.go [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+)
+
+// OTValue is used in the JetStream offset timeline bucket as the value for the given processor entity key.
+type OTValue struct {
+	Offset    int64
+	Watermark int64
+}
+
+// EncodeToBytes encodes a OTValue object into byte array.
+func (v OTValue) EncodeToBytes() ([]byte, error) {
```
Comment (KeranYang) [inline pkg/watermark/otbucket/otbucket.go:15]: Shall we add a simple unit test to encode then decode a OTValue and compare result?

Comment (chromevoid) [reply pkg/watermark/otbucket/otbucket.go:15]: That’s why it’s still a draft PR.

File: pkg/watermark/otbucket/otbucket.go (hunk 1 above)
Comment (vigith) [reply pkg/watermark/otbucket/otbucket.go:1]: just ot? it is anyways encapsulating OTValue

File: pkg/watermark/otbucket/otbucket.go [hunk 8]
```diff
@@ -0,0 +1,33 @@
+package otbucket
```
Comment (vigith) [inline pkg/watermark/otbucket/otbucket.go:1]:
```suggestion
// Package ot ....doc....
package otbucket
```

assuming you change the name to `ot`

File: pkg/watermark/publish/publisher.go [hunk 9]
```diff
@@ -140,26 +143,21 @@ func (p *publish) PublishWatermark(wm processor.Watermark, offset isb.Offset) {
...
-		}
-		if err != nil {
-			p.log.Errorw("Invalid epoch time string", zap.Error(err))
-			continue
-		}
-		if latestWatermark < epoch {
-			latestWatermark = epoch
-		}
+	var (
+		timeWatermark = time.UnixMilli(math.MinInt64)
```
Comment (vigith) [inline pkg/watermark/publish/publisher.go:147]: isn't -1 enough?

**Review by vigith — COMMENTED** [review]
thinking how can we do a comparison in before and after performance , any ideas?

**Comment by chromevoid** [comment]
> thinking how can we do a comparison in before and after performance , any ideas?

run the same pipeline which led us to this issue?

**Comment by vigith** [comment]
> run the same pipeline which led us to this issue?

It won't show any performance improvement. It will show that the KV entries are less.

By performance, i meant whether we can see any perceivable improvement is throughput, speed, CPU, etc.

File: pkg/watermark/ot/ot.go [hunk 10]
```diff
@@ -0,0 +1,34 @@
+// Package ot defines the ot value struct and how to encode to /decode from []byte
```
Comment (vigith) [inline pkg/watermark/ot/ot.go:1]:
```suggestion
// Package ot represents the offset-timeline pair and its corresponding encoder and decoder.
```

**Review by vigith — APPROVED** [review]

**Review by whynowy — APPROVED** [review]

**Pull request merged**